packages/loopring_v3/contracts/amm/libamm/AmmExitProcess.sol

```diff
@@ -79,93 +38,152 @@ library AmmExitProcess
-                    transfer.feeTokenID == ctx.tokens[i].tokenID &&
+                    transfer.feeTokenID == 0 &&
                     transfer.fee == 0,
                     "INVALID_TX_DATA"
                 );
 
-                // Replay protection when using a signature (otherwise the approved hash is cleared onchain)
-                if (signature.length > 0) {
-                    require(transfer.storageID == exit.storageIDs[i], "INVALID_TX_DATA");
-                }
-
-                // Now approve this transfer
                 transfer.validUntil = 0xffffffff;
                 bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
                 ctx.exchange.approveTransaction(address(this), txHash);
-
-                amount = transfer.amount;
-                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].sub(amount);
-            } else {
-                address token = ctx.tokens[i].addr;
-                // Make the amount available for withdrawing
-                S.userBalance[token][exit.owner] = S.userBalance[token][exit.owner].add(amount);
             }
+        }
 
-            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].sub(amount);
+        // Handle pool token
+        ctx.totalSupply = ctx.totalSupply.sub(exit.burnAmount);
+
+        if (burnFromLayer1) {
+            S.burn(address(this), exit.burnAmount);
+        } else {
+            _approvePoolTokenWithdrawal(ctx, exit.burnAmount, exit.owner, exit.burnStorageID, signature);
         }
+    }
+
+    function _approvePoolTokenWithdrawal(
+        AmmData.Context  memory  ctx,
+        uint                     amount,
+        address                  from,
+        uint32                   burnStorageID,
+        bytes            memory  signature
+        )
+        internal
+    {
+        // Check that the withdrawal in the block matches the expected withdrawal
+        WithdrawTransaction.Withdrawal memory withdrawal = ctx._block.readWithdrawal(ctx.txIdx++);
+        ctx.numTransactionsConsumed++;
+
+        // These fields are not read by readWithdrawal: to, extraData, minGas, validUntil
+        withdrawal.to = address(this);
+        withdrawal.extraData = new bytes(0);
+        withdrawal.minGas = 0;
+        withdrawal.validUntil = 0xffffffff;
 
-        S.removeUserBalance(address(this), exit.owner, exit.poolAmountIn);
-        S.burn(address(this), exit.poolAmountIn);
+        bytes32 expectedOnchainDataHash = WithdrawTransaction.hashOnchainData(
+            withdrawal.minGas,
+            withdrawal.to,
+            withdrawal.extraData
+        );
+
+
+        if (signature.length > 0) {
+            require(withdrawal.storageID == burnStorageID, "INVALID_STORAGE_ID_OR_REPLAY");
+        }
+
+        require(
+            withdrawal.withdrawalType == 1 &&
+            withdrawal.owner == from &&
+            withdrawal.accountID == 0 &&
```

Comment:
Will have to be ignored (which is fine), `accountID == 0`won't be valid (`0` is not a special value for any account). I guess it could also come from the exit request, but I don't much reason for that.